### ISSUE TITLE
How is dependency managed in kcc

### ISSUE THREAD

**Issue by yashsaini77** [GoogleCloudPlatform/k8s-config-connector#343]
I want to know that I want to create a trigger with cloud build to a git repo whenever we push a sample file to Git so Cloud Build should deploy a couple of YAML files to create these GCP resources with KCC.
- Folder
- Project 
- Service Account 
- Custom VPC
- Compute Engine, 
- GKE Cluster 

These resources should be created in order like for example folder should be created first then only a project can be created under that folder so how we can determine this kind of dependency? How we can determine that a VM and GKE Cluster shouldn't start creating it, before the Service account and Custom VPC has been created? 

Note: There is a folder named k8s in my Git repo which contains all the manifest files(YAML) for each of the resources (Project.yaml, Folder.yaml, Custom-vpc.yaml and so on).

**Comment by caieo** [comment]
Hi @yashsaini77, you don't need to determine the exact dependencies yourself, Config Connector will handle that. Config Connector will use eventual consistency to reconcile your resources to their desired state. This means that if  you have a project that relies on a folder to be created first, you can apply both at the same time and the project won't be ready until the folder is. You can read more about this on our [reconciliation concept documentation here ](https://cloud.google.com/config-connector/docs/concepts/reconciliation).

**Comment by yashsaini77** [comment]
@caieo Let's say I want to make a trigger on cloud build. I want that in my git repo whenever someone pushes some file so cloud build should deploy all the infra resources(YAML files available in k8s folder) to the GKE cluster. In this case, we don't know if all the files will be applied at the same time or not? So are you sure that even though whichever files get applied first, will create all the resources the desired way without any issue?

 To get started with Cloud Build, I create a cloud build config file like this

#step 1
- name: 'gcr.io/cloud-builders/kubectl'
  args: ['apply', '-f', 'k8s/']
  env:
  - 'CLOUDSDK_COMPUTE_ZONE=<cluster-zone>'
  - 'CLOUDSDK_CONTAINER_CLUSTER=<cluster-name>'

Once this file will be run through cloud build (after the trigger executes) it wont know which YAML file to deploy first ?

**Comment by caieo** [comment]
@yashsaini77 Yes, regardless of when or which order the files are applied, the dependencies will eventually be reconciled correctly. Resources that are dependent on other resources will be marked as 'updating' until their dependencies are ready.

**Comment by caieo** [comment]
@yashsaini77, sorry, I misled you about our Folder resource -- I saw your comment in #104 and remembered that Folders have a generated ID as their name and must be created before being able to reference it in other resources. We will update #104 when we support the case mentioned there.

**Comment by yashsaini77** [comment]
Hi **@caieo**. no problem so you saying this is not possible to do at this time what is the solution for it ? 
and apart from folder, how about the other resources that I have mentioned in my issue upwards. In those resources you think anything won't clash ?

**Comment by caieo** [comment]
@yashsaini77 , As per #104, it is currently not possible to reference the folderID as part of the folder-id annotation. 

The rest of your resources should eventually reconcile into your desired state without issue.

**Comment by ldanielmadariaga** [comment]
This item has been open for >12 months. As part of scheduled repository maintenance, this Issue/PR will be closed. If this is still a critical ask, please re-open and work with your Google Cloud Account team to prioritize it with GCP KCC Product Management.
